File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.js
const DISCOUNT_RATE = [10, 20, 30, 40];

const connectDiscountRate = (results, length) => {
    if (results.length >= length) {
        return [results]
    }
    const ret = []
    for (const rate of DISCOUNT_RATE) {
        ret.push(...connectDiscountRate([...results, rate], length))
    }
    return ret
}

const zip = (arr1, arr2) => {
    const min = Math.min(arr1.length, arr2.length);
    return Array.from({length: min}, (_, i) => [arr1[i], arr2[i]]);
}

const discount = (price, rate) => price - (price * rate / 100);
const shouldJoinEmoticonPlus = (total, max) => total >= max;
const getTotal = (rates) => (emoticons) => ([maxRate, maxPrice]) => zip(emoticons, rates)
    .reduce((acc, [emoticon, rate]) => 
            rate >= maxRate 
                ? acc + discount(emoticon, rate) 
                : acc, 
            0)

function solution(users, emoticons) {
    const ratePerm = connectDiscountRate([], emoticons.length);
    
    const maxPrices = ratePerm.map(rates => {
        const totalPrices = users.map(user => {
            const total = getTotal(rates)(emoticons)(user);
            const shouldJoined = shouldJoinEmoticonPlus(total, user[1])
            
            return {
                total: shouldJoined ? 0 : total,
                joined: shouldJoined ? 1 : 0,
            }
            
           
        }).reduce((acc, cur) => {
            const total = acc.total + cur.total;
            const joined = acc.joined + cur.joined
            return {total, joined, rates};
        }, {total: 0, joined: 0})

        return totalPrices;
    })
    const [answer] = maxPrices.sort((a, b) => b.joined - a.joined || b.total - a.total);
    
    return [answer.joined, answer.total];
}

File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.py
def get_emoticon_cost(emoticon_cost, discount_rate):
    return emoticon_cost * (1 - discount_rate * 0.01)

def get_total_pay(user, emoticons_with_rate):
    pay = 0
    for emoticon in emoticons_with_rate:
        if user["rate"] <= emoticon["rate"]:
            pay += get_emoticon_cost(emoticon["cost"], emoticon["rate"])
    return pay

def is_joined(user, total_pay):
    return total_pay >= user["cost"]

def create_user(user):
    rate, cost = user
    return {
        "rate": rate,
        "cost": cost
    }
    
def create_emoticon(emoticon, rate):
    return {
        "cost": emoticon, 
        "rate": rate
    }
    
def get_result(users, emoticons_with_rate):
    joined = 0
    cost = 0
    for user in users:
        total = get_total_pay(user, emoticons_with_rate)
        if is_joined(user, total):
            joined += 1
        else:
            cost += total
    return [joined, cost]
    
def recur(emoticons, rates, users, results):
    if len(rates) >= len(emoticons):
        emoticons_with_rate = [create_emoticon(e, r) for e, r in zip(emoticons, rates)]
        results.append(get_result(users, emoticons_with_rate))
        return
    
    for rate in [10, 20, 30, 40]:
        recur(emoticons, rates + [rate], users, results)
    
def solution(users, emoticons):
    discount_rates = [10, 20, 30, 40]
    results = []
    users = [create_user(user) for user in users]
    recur(emoticons, [], users, results)
    results.sort(key = lambda x: (-x[0], -x[1]))
    return results[0]